README.rst
.. image:: https://img.shields.io/badge/License-MIT-green.svg
    :alt: domain-availability-py license
    :target: https://opensource.org/licenses/MIT

.. image:: https://img.shields.io/pypi/v/domain-availability.svg
    :alt: domain-availability-py release
    :target: https://pypi.org/project/domain-availability

.. image:: https://github.com/whois-api-llc/domain-availability-py/workflows/Build/badge.svg
    :alt: domain-availability-py build
    :target: https://github.com/whois-api-llc/domain-availability-py/actions

========
Overview
========

The client library for
`Domain Availability API <https://domain-availability.whoisxmlapi.com/>`_
in Python language.

The minimum Python version is 3.6.

Installation
============

.. code-block:: shell

    pip install domain-availability

Examples
========

Full API documentation available `here <https://domain-availability.whoisxmlapi.com/api/documentation/making-requests>`_

Create a new client
-------------------

.. code-block:: python

    from domainavailability import *

    client = Client('Your API key')

Make basic requests
-------------------

.. code-block:: python

    # Get parsed records as a model instance.
    result = client.data('whoisxmlapi.com')
    print(result.domain_availability)
    if result.is_available() is True:
        print('This domain name is available.')

    # Get raw API response
    raw_result = client.raw_data('whoisxmlapi.com')

Advanced usage
-------------------

Extra request parameters

.. code-block:: python

    result = client.data(
        'whoisxmlapi.com',
        mode=Client.DNS_AND_WHOIS_MODE,
        credits_type=Client.DOMAIN_AVAILABILITY_CREDITS)

    raw_result = client.raw_data(
        'whoisxmlapi.com',
        mode=Client.DNS_ONLY_MODE,
        credits_type=Client.WHOIS_CREDITS,
        output_format=Client.XML_FORMAT)


CHANGELOG.rst

Changelog
=========

1.0.0 (2021-04-26)
------------------

* First release

tests/model_test.py
import unittest
from json import loads
from domainavailability import DomainInfo, ErrorMessage


_json_response_ok_unavailable = '''{
   "DomainInfo": {
      "domainAvailability": "UNAVAILABLE",
      "domainName": "google.com"
   }
}'''

_json_response_ok_available = '''{
    "DomainInfo": {
        "domainAvailability": "AVAILABLE",
        "domainName": "dsajkdhksajhdkjsahdkjshakjdhsakdhksajhdsakdsadjksahdkjas.com"
    }
}'''

_json_response_error = '''{
    "code": 403,
    "messages": "Access restricted. Check credits balance or enter the correct API key."
}'''


class TestModel(unittest.TestCase):

    def test_response_parsing(self):
        response = loads(_json_response_ok_available)
        parsed = DomainInfo(response['DomainInfo'])
        assert parsed.domain_name == 'dsajkdhksajhdkjsahdkjshakjdhsakdhksajhdsakdsadjksahdkjas.com'
        assert parsed.domain_availability == 'AVAILABLE'
        assert parsed.is_available()

        response = loads(_json_response_ok_unavailable)
        parsed = DomainInfo(response['DomainInfo'])
        assert parsed.domain_name == 'google.com'
        assert parsed.domain_availability == 'UNAVAILABLE'
        assert parsed.is_available() is False

    def test_error_parsing(self):
        error = loads(_json_response_error)
        parsed_error = ErrorMessage(error)
        assert parsed_error.code == 403
        assert parsed_error.message == 'Access restricted. Check credits balance or enter the correct API key.'


tests/client_test.py
import os
import unittest
from domainavailability import Client
from domainavailability import ParameterError, ApiAuthError


class TestClient(unittest.TestCase):
    """
    Final integration tests without mocks.

    Active API_KEY is required.
    """
    def test_get_correct_data(self):
        client = Client(os.getenv("API_KEY"))

        domain = 'whoisxmlapi.com'
        result = client.data(domain)
        assert result.domain_name == domain
        assert result.domain_availability == 'UNAVAILABLE'
        assert not result.is_available()
        result = client.data(
            domain,
            mode=Client.DNS_AND_WHOIS_MODE,
            credits_type=Client.WHOIS_CREDITS)
        assert not result.is_available()

    def test_get_raw_data(self):
        client = Client(os.getenv("API_KEY"))
        raw = client.raw_data(
            'whoisxmlapi.com', output_format=Client.XML_FORMAT)
        assert raw.startswith('<?xml')

    def test_get_auth_error(self):
        client = Client('at_00000000000000000000000000000')
        self.assertRaises(ApiAuthError, client.data, 'whoisxmlapi.com')
        try:
            client.data('whoisxmlapi.com')
        except ApiAuthError as error:
            parsed = error.parsed_message
            assert parsed.code == 403

    def test_get_parameter_error(self):
        client = Client(os.getenv('API_KEY'))
        domain = 'whoisxmlapi.com'
        self.assertRaises(ParameterError, client.data, 'incorrect-domain')

        with self.assertRaises(ParameterError):
            client.raw_data(domain, output_format='pdf')
        with self.assertRaises(ParameterError):
            client.data(domain, mode='full')
        with self.assertRaises(ParameterError):
            client.data(domain, credits_type='drs')


if __name__ == '__main__':
    unittest.main()


src/domainavailability/net/http.py
from requests import request
from ..exceptions.error import ApiAuthError, HttpApiError, BadRequestError
import logging


class ApiRequester:
    __logger = logging.getLogger("domain-availability")
    __connect_timeout = 5
    _base_url: str
    _timeout: float

    def __init__(self, **kwargs):
        """

        :param kwargs: Supported parameters:
        - base_url: (optional) API endpoint URL; str
        - timeout: (optional) API call timeout in seconds; float
        """
        self._base_url = ''
        self.timeout = 30

        if 'base_url' in kwargs:
            self.base_url = kwargs['base_url']
        if 'timeout' in kwargs:
            self.timeout = kwargs['timeout']

    @property
    def base_url(self) -> str:
        return self._base_url

    @base_url.setter
    def base_url(self, url: str):
        if url is None or len(url) <= 8 or not url.startswith('http'):
            raise ValueError("Invalid URL specified.")
        self._base_url = url

    @property
    def timeout(self) -> float:
        """API call timeout in seconds"""
        return self._timeout

    @timeout.setter
    def timeout(self, value: float):
        """API call timeout in seconds"""
        if value is not None and 1 <= value <= 60:
            self._timeout = value
        else:
            raise ValueError("Timeout value should be in [1, 60]")

    def get(self, payload: dict):
        response = request(
            "GET",
            self.base_url,
            params=payload,
            timeout=(ApiRequester.__connect_timeout, self.timeout)
        )

        if 200 <= response.status_code < 300:
            return response.text

        if response.status_code in [401, 402, 403]:
            raise ApiAuthError(response.text)

        if response.status_code in [400, 422]:
            raise BadRequestError(response.text)

        if response.status_code >= 300:
            raise HttpApiError(response.text)


src/domainavailability/models/response.py
import copy
from .base import BaseModel


def _string_value(values: dict, key: str) -> str:
    if key in values:
        return str(values[key])
    return ''


def _int_value(values: dict, key: str) -> int:
    if key in values:
        return int(values[key])
    return 0


def _list_value(values: dict, key: str) -> list:
    if key in values and type(values[key]) is list:
        return copy.deepcopy(values[key])
    return []


def _list_of_objects(values: dict, key: str, classname: str) -> list:
    r = []
    if key in values and type(values[key]) is list:
        r = [globals()[classname](x) for x in values[key]]
    return r


class DomainInfo(BaseModel):
    domain_availability: str
    domain_name: str

    __AVAILABLE = 'AVAILABLE'
    __UNAVAILABLE = 'UNAVAILABLE'

    def __init__(self, values):
        super().__init__()

        self.domain_availability = ''
        self.domain_name = ''

        if values is not None:
            self.domain_name = _string_value(values, 'domainName')
            self.domain_availability = _string_value(
                values, 'domainAvailability')

    def is_available(self) -> bool or None:
        """
        :returns: True if available, else False.
            If there is no value will return None.
        """
        if self.domain_availability == '':
            return None

        return self.domain_availability == DomainInfo.__AVAILABLE


class ErrorMessage(BaseModel):
    code: int
    message: str

    def __init__(self, values):
        super().__init__()

        self.int = 0
        self.message = ''

        if values is not None:
            self.code = _int_value(values, 'code')
            self.message = _string_value(values, 'messages')


src/domainavailability/exceptions/error.py
from json import loads
from ..models.response import ErrorMessage


class DomainAvailabilityApiError(Exception):
    def __init__(self, message):
        self.message = message

    @property
    def message(self):
        return self._message

    @message.setter
    def message(self, message):
        self._message = message

    def __str__(self):
        return str(self.__dict__)


class ParameterError(DomainAvailabilityApiError):
    pass


class EmptyApiKeyError(DomainAvailabilityApiError):
    pass


class ResponseError(DomainAvailabilityApiError):
    def __init__(self, message):
        self.message = message
        self._parsed_message = None
        try:
            parsed = loads(message)
            self.parsed_message = ErrorMessage(parsed)
        except Exception:
            pass

    @property
    def parsed_message(self):
        return self._parsed_message

    @parsed_message.setter
    def parsed_message(self, pm):
        self._parsed_message = pm


class UnparsableApiResponseError(DomainAvailabilityApiError):
    def __init__(self, message, origin_error):
        self.message = message
        self.original_error = origin_error

    @property
    def original_error(self):
        return self._original_error

    @original_error.setter
    def original_error(self, oe):
        self._original_error = oe


class ApiAuthError(ResponseError):
    pass


class BadRequestError(ResponseError):
    pass


class HttpApiError(DomainAvailabilityApiError):
    pass


src/domainavailability/client.py
from json import loads, JSONDecodeError
import re

from .net.http import ApiRequester
from .models.response import DomainInfo
from .exceptions.error import ParameterError, EmptyApiKeyError, \
    UnparsableApiResponseError


class Client:
    __default_url = "https://domain-availability.whoisxmlapi.com/api/v1"
    __parsable_format = 'json'
    _api_requester: ApiRequester or None
    _api_key: str
    _mode: str
    _credits_type: str

    _re_api_key = re.compile(r'^at_[a-z0-9]{29}$', re.IGNORECASE)
    _re_output_format = re.compile(r'^(json)|(xml)$', re.IGNORECASE)
    _re_domain_name = re.compile(
        r'^(?:[0-9a-z_](?:[0-9a-z-_]{0,62}(?<=[0-9a-z-_])[0-9a-z_])?\.)+'
        + r'[0-9a-z][0-9a-z-]{0,62}[a-z0-9]$', re.IGNORECASE)

    DNS_AND_WHOIS_MODE = 'DNS_AND_WHOIS'
    DNS_ONLY_MODE = 'DNS_ONLY'

    WHOIS_CREDITS = 'WHOIS'
    DOMAIN_AVAILABILITY_CREDITS = 'DA'

    JSON_FORMAT = 'JSON'
    XML_FORMAT = 'XML'

    def __init__(self, api_key: str, **kwargs):
        """
        :param api_key: str: Your API key.
        :param credits_type: :str: (optional)
        :param mode: :str: (optional)
        :param base_url: str: (optional) API endpoint URL.
        :param timeout: float: (optional) API call timeout in seconds
        """
        self._api_key = ''
        self._mode = Client.DNS_AND_WHOIS_MODE
        self._credits_type = Client.WHOIS_CREDITS

        self.api_key = api_key

        if 'base_url' not in kwargs:
            kwargs['base_url'] = Client.__default_url

        if 'credits_type' in kwargs:
            self.credits_type = kwargs['credits_type']
        if 'mode' in kwargs:
            self.mode = kwargs['mode']

        self.api_requester = ApiRequester(**kwargs)

    @property
    def api_key(self) -> str:
        return self._api_key

    @api_key.setter
    def api_key(self, value: str):
        self._api_key = Client._validate_api_key(value)

    @property
    def api_requester(self) -> ApiRequester or None:
        return self._api_requester

    @api_requester.setter
    def api_requester(self, value: ApiRequester):
        self._api_requester = value

    @property
    def base_url(self) -> str:
        return self._api_requester.base_url

    @base_url.setter
    def base_url(self, value: str or None):
        if value is None:
            self._api_requester.base_url = Client.__default_url
        else:
            self._api_requester.base_url = value

    @property
    def credits_type(self) -> str:
        return self._credits_type

    @credits_type.setter
    def credits_type(self, value: str):
        try:
            self._credits_type = Client._validate_credits_type(value)
        except ParameterError as e:
            raise ValueError(e.message)

    @property
    def mode(self) -> str:
        return self._mode

    @mode.setter
    def mode(self, value: str):
        try:
            self._mode = Client._validate_mode(value)
        except ParameterError as e:
            raise ValueError(e.message)

    @property
    def timeout(self) -> float:
        return self._api_requester.timeout

    @timeout.setter
    def timeout(self, value: float):
        self._api_requester.timeout = value

    def data(self, domain: str, **kwargs) -> DomainInfo:
        """

        :param domain: - :str|IPv4Address: - the domain name
        :param mode: (optional) :str: Mode - use constants DNS_ONLY_MODE
            and DNS_AND_WHOIS_MODE
        :param credits_type: (optional) :str: - use constants
            WHOIS_CREDITS and DOMAIN_AVAILABILITY_CREDITS
        :return: DomainInfo
        :raises:
        - base class is DomainAvailabilityApiError
          - EmptyApiKeyError
          - ResponseError -- response contains an error message
          - ApiAuthError -- Server returned 401, 402 or 403 HTTP code
          - BadRequestError - Server returned 400 or 422 HTTP code
          - HttpApiError -- HTTP code >= 300 and not equal to above codes from
          - UnparsableApiResponseError -- the response couldn't be parsed
          - ParameterError -- invalid parameter's value
        - ConnectionError
        """

        kwargs['output_format'] = Client.__parsable_format

        response = self.raw_data(domain, **kwargs)
        try:
            parsed = loads(str(response))
            if 'DomainInfo' in parsed:
                return DomainInfo(parsed['DomainInfo'])
            raise UnparsableApiResponseError(
                "Could not find the correct root element.", None)
        except JSONDecodeError as error:
            raise UnparsableApiResponseError("Could not parse API response", error)

    def raw_data(self, domain: str, **kwargs) -> str:
        """

        :param domain: - :str|IPv4Address: - the domain name
        :param mode: (optional) :str: Mode - use constants DNS_ONLY_MODE
            and DNS_AND_WHOIS_MODE
        :param credits_type: (optional) :str: - use constants
            WHOIS_CREDITS and DOMAIN_AVAILABILITY_CREDITS
        :param output_format: - (optional) :str: use constants JSON_FORMAT
            and XML_FORMAT
        :return: str
        :raises
        - base class is domainavailability.exceptions.ReverseIpApiError
          - ResponseError -- response contains an error message
          - ApiAuthError -- Server returned 401, 402 or 403 HTTP code
          - BadRequestError - Server returned 400 or 422 HTTP code
          - HttpApiError -- HTTP code >= 300 and not equal to above codes
          - ParameterError -- invalid parameter's value
        - ConnectionError
        """
        if self.api_key == '':
            raise EmptyApiKeyError('')

        domain = Client._validate_domain_name(domain)

        if 'credits_type' in kwargs:
            credits_type = Client._validate_credits_type(
                kwargs['credits_type'])
        else:
            credits_type = self.credits_type

        if 'output_format' in kwargs:
            output_format = Client._validate_output_format(
                kwargs['output_format'])
        else:
            output_format = Client.__parsable_format

        if 'mode' in kwargs:
            mode = Client._validate_mode(kwargs['mode'])
        else:
            mode = self.mode

        return self._api_requester.get(self._build_payload(
            domain, credits_type, mode, output_format
        ))

    @staticmethod
    def _validate_api_key(api_key):
        if Client._re_api_key.search(
                str(api_key)
        ) is not None:
            return str(api_key)
        else:
            raise ParameterError("Invalid API key format.")

    @staticmethod
    def _validate_mode(value: str):
        if value.upper() not in [Client.DNS_AND_WHOIS_MODE,
                                 Client.DNS_ONLY_MODE]:
            raise ParameterError('Incorrect value.')
        return value.upper()

    @staticmethod
    def _validate_credits_type(value: str):
        if value.upper() not in [Client.WHOIS_CREDITS,
                                 Client.DOMAIN_AVAILABILITY_CREDITS]:
            raise ParameterError('Incorrect value.')
        return value.upper()

    @staticmethod
    def _validate_domain_name(domain):
        if domain is not None and len(str(domain)) > 0:
            if Client._re_domain_name.match(domain) is not None:
                return domain
        raise ParameterError('Invalid domain name.')

    @staticmethod
    def _validate_output_format(_format):
        if Client._re_output_format.search(str(_format)) is not None:
            return str(_format)
        else:
            raise ParameterError(
                "Output format should be either JSON or XML.")

    def _build_payload(self, domain, credits_type, mode, output_format):
        return {
            'apiKey': self.api_key,
            'domainName': domain,
            'credits': credits_type,
            'mode': mode,
            'outputFormat': output_format
        }
